<?xml version="1.0" encoding="ISO-8859-15"?>

<launch>

  <param name="/use_sim_time" value="true"/>

  <!-- Play rosbag -->
  <arg name="path" default="$(find mel_amcl)/bag_data/local_odom/" />
  <arg name="file_name" default="local_2019-07-15-13-47-10.bag" />
  <node pkg="rosbag" type="play" name="player" output="screen" args="--clock $(arg path)$(arg file_name) -r 2"/>


  <!-- rviz -->
  <node pkg="rviz" type="rviz" name="rviz"
          args="-d $(find mel_amcl)/rviz/odom_tuning.rviz"/>

  <!-- load parameters for eKFs: odom & odom_imu  -->
  <rosparam command="load" file="$(find mel_amcl)/config/odom_ekf_test.yaml" />

  <!-- eKF odom only -->
  <node pkg="robot_localization" type="ekf_localization_node" name="ekf_odom" clear_params="true">
    <remap from="odometry/filtered" to="odometry/odom_only"/>
  </node>

  <!-- eKF odom & um7 imu fusion only -->
  <node pkg="robot_localization" type="ekf_localization_node" name="ekf_odom_imu" clear_params="true">
    <remap from="odometry/filtered" to="odometry/odom_imu"/>
  </node>


  <!-- eKF odom & um7 imu & laser fusion only -->
  <node pkg="robot_localization" type="ekf_localization_node" name="ekf_odom_imu_laser" clear_params="true">
    <remap from="odometry/filtered" to="odometry/odom_imu_laser"/>
  </node>

  <!-- eKF odom & xsens imu & laser fusion only -->
  <node pkg="robot_localization" type="ekf_localization_node" name="ekf_odom_imu_xsens" clear_params="true">
    <remap from="odometry/filtered" to="odometry/odom_imu_xsens"/>
  </node>

  <!-- Laser scan based odometry -->
  <arg name="publish_covariance" default="true"/>
  <group if="$(arg publish_covariance)">
    <param name="laser_scan_matcher_node/do_compute_covariance" value="1"/>
    <param name="laser_scan_matcher_node/publish_pose_with_covariance" value="true"/>
    <param name="laser_scan_matcher_node/publish_pose_with_covariance_stamped" value="true"/>
  </group>
  <node pkg="laser_scan_matcher" type="laser_scan_matcher_node"
    name="laser_scan_matcher_node" output="screen">
    <param name="max_iterations" value="20"/>
    <param name="fixed_frame" value="odom"/>
    <param name="publish_tf" value="true"/>
    <remap from="pose_with_covariance_stamped" to="odometry/laser_scan_match"/>
  </node>

  <node pkg="hector_trajectory_server" type="hector_trajectory_server" name="hector_trajectory_server" output="screen">
    <param name="target_frame_name" type="string" value="odom" />
    <param name="source_frame_name" type="string" value="base_link" />
    <param name="trajectory_update_rate" type="double" value="4" />
    <param name="trajectory_publish_rate" type="double" value="4" />
  </node>


  <!-- eKF for the gps in map frame, to use output, set use_gps_odom arg in mel_amcl to true --> 
  <!-- <node pkg="robot_localization" type="ekf_localization_node" name="ekf_gps" clear_params="true">
    <remap from="odometry/filtered" to="odometry/gps"/>
  </node> -->


</launch>
